examples/discovery.js

```diff
@@ -0,0 +1,217 @@
+var TChannelAsThrift = require('tchannel/as/thrift');
+var Ringpop = require('../index.js');
+var RingpopHandler = require('../ringpop-handler.js');
+
+var thriftSource = fs.readFileSync(path.join(__dirname, 'hyperbahn.thrift'), 'utf8');
+
+/* Instructions:
+ * 1) run Hyperbahn with at least one node: 127.0.0.1:21300
+ *
+ * 2) node discovery.js
+ *    You will observe the member joining process. 
+ *
+ * Issue:
+ *    If you run this example more than once without restarting Hyperbahn,
+ *    ringpop will refuses to join, since the host list returned by discovery API
+ *    has dead hostports from previous runs.
+ *
+ */
+
+function App(options) {
+    if (!(this instanceof App)) {
+        return new App(options);
+    }
+
+    var self = this;
+
+    self.thrift = new TChannelAsThrift({source: thriftSource});
+
+    self.name = options.name;
+
+    self.channel = TChannel();
+}
+
+App.prototype.setupRingpop = function setupRingpop() {
+    var self = this;
+    self.ringpop = Ringpop({
+        app: 'app',
+        hostPort: self.channel.address().address + ':' + self.channel.address().port,
+        channel: self.channel.makeSubChannel({
+            serviceName: 'ringpop'
+        })
+    });
+
+    self.appChannel = self.channel.makeSubChannel({
+        serviceName: 'app'
+    });
+    self.appChannel.handler = RingpopHandler({
+        channel: self.appChannel,
+        ringpop: self.ringpop,
+        logger: self.channel.logger,
+        realHandler: self.appChannel.handler
+    });
+
+    self.appChannel.register('hello', function hello(req, res, arg2, arg3) {
+        res.headers.as = 'raw';
+        res.sendOk('', 'hello from ' + self.name + ' for ' + req.headers.sk);
+    });
+
+    self.ringpop.setupChannel();
+}
+
+App.prototype.discover = function discover(cb) {
+    var self = this;
+    self.hyperbahnChannel = self.hyperbahnChannel || self.hyperbahnClient.getClientChannel({
+        serviceName: 'hyperbahn'
+    });
+
+    var request = self.hyperbahnChannel.request({
+        headers: {
+            cn: 'app'
+        },
+        serviceName: 'hyperbahn',
+        hasNoParent: true
+    });
+    self.thrift.send(request,
+        'Hyperbahn::discover',
+        null,
+        {
+            query: {
+                serviceName: 'app'
+            }
+        },
+        onResponse
+    );
+
+    function onResponse(err, res) {
+        if (err || !res.ok) {
+            console.log(res);
+            console.log('call to discovery API failed', {
+                error: err,
+                body: res.body
+            });
+            cb([]);
+            return;
+        }
+
+        var hosts = [];
+        for (var i = 0; i < res.body.peers.length; i++) {
+            hosts.push(covertHost(res.body.peers[i]));
+        }
+
+        cb(hosts);
+    }
+};
+
+App.prototype.bootstrap = function bootstrap(cb) {
+    var self = this;
+
+    // listen
+    self.channel.listen(0, '127.0.0.1', onListen);
```

Comment:
Listening on `0` is not going to jive well with Ringpop, as it is currently implemented today. On restart, the node that previously identified itself as host:port `X:Y` needs to be able to preserve that identity. This is not guaranteed in the port `0` model as far as I understand unless Ringpop builds in a higher-level identification mechanism that is different than its listening address. I think that's the first problem to solve, then the port 0 thing will be not a problem.